### PR TITLE
Optimized roc_auc_scorer speed by 5x

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*
- Optimized roc_auc_scorer to be 5x faster with no downside on binary problems.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by tonyhoo** [comment]
Can you include some benchmark results for the improvements we observed from the customized ROC impl?

File: core/src/autogluon/core/metrics/classification_metrics.py [hunk 1]
```diff
@@ -366,3 +368,70 @@ def quadratic_kappa(y_true, y_pred):
...
+    if kwargs:
+        return sklearn.metrics.roc_auc_score(y_true, y_score, **kwargs)
+
+    desc_score_indices = np.argsort(y_score, kind="mergesort")[::-1]
+    y_score = y_score[desc_score_indices]
+    y_true = y_true[desc_score_indices]
+
+    # keep only indices that have different values to speed up future computation
+    distinct_value_indices = np.where(np.diff(y_score))[0]
+    threshold_idxs = np.r_[distinct_value_indices, y_true.size - 1]
```
Comment (gradientsky) [inline core/src/autogluon/core/metrics/classification_metrics.py:407]: the doc says `array-like of int` there's no `.size` on int

Comment (gradientsky) [reply core/src/autogluon/core/metrics/classification_metrics.py:407]: nit: please add quick comment on what's going on here. `np.r_` is not commonly used API and it requires maintainers to do additional exploration when reading this code.

Nice trick, BTW

File: core/src/autogluon/core/metrics/classification_metrics.py [hunk 2]
```diff
@@ -366,3 +368,70 @@ def quadratic_kappa(y_true, y_pred):
...
+    roc_auc_score : float
+        The roc_auc_score in higher_is_better format.
+    """
+    if isinstance(y_true, pd.Series):
+        y_true = y_true.values
+    if isinstance(y_score, pd.Series):
+        y_score = y_score.values
+    if kwargs:
+        return sklearn.metrics.roc_auc_score(y_true, y_score, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply core/src/autogluon/core/metrics/classification_metrics.py:400]: defensive boundary condition check: `[]` (no elements) input

File: core/tests/unittests/metrics/test_classification_metrics.py [hunk 3]
```diff
@@ -192,6 +192,18 @@ def test_log_loss_with_sklearn(gt, probs):
     np.testing.assert_allclose(ag_loss_as_sklearn, sklearn_log_loss)
 
 
+def test_roc_auc_score_with_sklearn():
```
Comment (gradientsky) [inline core/tests/unittests/metrics/test_classification_metrics.py:195]: please add empty input `[]` test case and one-value input (int+float check): y_true=1, y_score=0.9

File: core/src/autogluon/core/metrics/classification_metrics.py (hunk 1 above)
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:407]: "the doc says array-like of int there's no .size on int"

Assumed to be numpy array

File: core/src/autogluon/core/metrics/classification_metrics.py [hunk 4]
```diff
@@ -366,3 +368,70 @@ def quadratic_kappa(y_true, y_pred):
...
+    desc_score_indices = np.argsort(y_score, kind="mergesort")[::-1]
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:401]: No meaningful difference, sklearn uses mergesort in this fashion so I just keep same. I tested both and it isn't a notice-able difference.

File: core/src/autogluon/core/metrics/classification_metrics.py [hunk 5]
```diff
@@ -366,3 +368,70 @@ def quadratic_kappa(y_true, y_pred):
...
+    **kwargs :
+        Any additional arguments. If not empty, will fallback to sklearn's implementation
+
+    Returns
+    -------
+    roc_auc_score : float
+        The roc_auc_score in higher_is_better format.
+    """
+    if isinstance(y_true, pd.Series):
+        y_true = y_true.values
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:395]: Thanks for checking!

Benchmark result:

```
0.15ms -> .values
0.20ms -> .to_numpy()
0.25ms -> .array
0.77ms -> sklearn native
```

Due to benchmark result I will continue using `.values` unless there is a functional gap.

File: core/src/autogluon/core/metrics/classification_metrics.py (hunk 1 above)
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:407]: Updating types to clarify

File: core/src/autogluon/core/metrics/classification_metrics.py [hunk 6]
```diff
@@ -366,3 +368,70 @@ def quadratic_kappa(y_true, y_pred):
...
+        y_score = y_score.values
+    if kwargs:
+        return sklearn.metrics.roc_auc_score(y_true, y_score, **kwargs)
+
+    desc_score_indices = np.argsort(y_score, kind="mergesort")[::-1]
+    y_score = y_score[desc_score_indices]
+    y_true = y_true[desc_score_indices]
+
+    # keep only indices that have different values to speed up future computation
+    distinct_value_indices = np.where(np.diff(y_score))[0]
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:406]: I named the variable aligning with sklearn: https://github.com/scikit-learn/scikit-learn/blob/f3f51f9b611bf873bd5836748647221480071a87/sklearn/metrics/_ranking.py#L774

I can change if you'd like, but then our naming will diverge from sklearn

File: core/src/autogluon/core/metrics/classification_metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:400]: The speedups come in from avoiding these checks. No need because we guarantee it is correct from how we access this in AutoGluon.

**Comment by Innixma** [comment]
> Good work with amazing results! Can you include some benchmark results for the improvements we observed from the customized ROC impl?

Definitely!

```python3
import numpy as np
import sklearn
from autogluon.core.metrics import roc_auc, make_scorer
import time

roc_auc_sklearn = make_scorer('roc_auc',
                      sklearn.metrics.roc_auc_score,
                      greater_is_better=True,
                      needs_threshold=True)

for sample_size in [
    10,
    100,
    1000,
    10000,
    100000,
]:
    y_true = np.random.randint(low=0, high=2, size=sample_size)
    y_score = np.random.random_sample(sample_size)

    print(f'### sample_size={sample_size} ###')
    ts = time.time()
    for i in range(1000):
        expected_score = roc_auc_sklearn(y_true, y_score)
    te = time.time()
    print(f'Sklearn: {round(te-ts, 2)}ms')

    ts = time.time()
    for i in range(1000):
        actual_score = roc_auc(y_true, y_score)
    te = time.time()
    print(f'Fast: {round(te-ts, 2)}ms')
    print(f'')
    assert expected_score == actual_score

```

Out:

```
### sample_size=10 ###
Sklearn: 0.71ms
Fast: 0.14ms
### sample_size=100 ###
Sklearn: 0.77ms
Fast: 0.15ms
### sample_size=1000 ###
Sklearn: 0.97ms
Fast: 0.26ms
### sample_size=10000 ###
Sklearn: 3.7ms
Fast: 1.36ms
### sample_size=100000 ###
Sklearn: 27.27ms
Fast: 15.74ms

```

File: core/tests/unittests/metrics/test_classification_metrics.py (hunk 3 above)
Comment (Innixma) [reply core/tests/unittests/metrics/test_classification_metrics.py:195]: Added + fixed edge case difference in output (returning NaN instead of exception), thanks for highlighting

File: core/src/autogluon/core/metrics/classification_metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:400]: Added empty check. dim check is fine because we have separate metrics for multiclass roc_auc

**Review by gradientsky — APPROVED** [review]

**Review by gidler — APPROVED** [review]
